units/event_send.go

```diff
@@ -0,0 +1,426 @@
+package units
+
+import (
+	"bytes"
+	"context"
+	"crypto/hmac"
+	"crypto/sha256"
+	"encoding/hex"
+	"fmt"
+	"net/http"
+	"net/url"
+	"strings"
+	"time"
+
+	"github.com/evergreen-ci/evergreen"
+	"github.com/evergreen-ci/evergreen/model/event"
+	"github.com/evergreen-ci/evergreen/model/notification"
+	"github.com/evergreen-ci/evergreen/util"
+	"github.com/mongodb/amboy"
+	"github.com/mongodb/amboy/dependency"
+	"github.com/mongodb/amboy/job"
+	"github.com/mongodb/amboy/registry"
+	"github.com/mongodb/grip"
+	"github.com/mongodb/grip/level"
+	"github.com/mongodb/grip/message"
+	"github.com/mongodb/grip/send"
+	"github.com/mongodb/grip/sometimes"
+	"github.com/pkg/errors"
+)
+
+const (
+	eventNotificationJobName = "event-send"
+
+	evergreenWebhookTimeout       = 5 * time.Second
+	evergreenNotificationIDHeader = "X-Evergreen-Notification-ID"
+	evergreenHMACHeader           = "X-Evergreen-Signature"
+)
+
+func init() {
+	registry.AddJobType(eventNotificationJobName, func() amboy.Job { return makeEventNotificationJob() })
+}
+
+type eventNotificationJob struct {
+	job.Base `bson:"job_base" json:"job_base" yaml:"job_base"`
+	settings *evergreen.Settings
+
+	NotificationID string `bson:"notification_id" json:"notification_id" yaml:"notification_id"`
+}
+
+func makeEventNotificationJob() *eventNotificationJob {
+	j := &eventNotificationJob{
+		Base: job.Base{
+			JobType: amboy.JobType{
+				Name:    eventNotificationJobName,
+				Version: 0,
+			},
+		},
+	}
+	j.SetDependency(dependency.NewAlways())
+
+	return j
+}
+
+func newEventNotificationJob(id string) amboy.Job {
+	j := makeEventNotificationJob()
+	j.NotificationID = id
+
+	j.SetID(fmt.Sprintf("%s:%s", eventNotificationJobName, id))
+	return j
+}
+
+func (j *eventNotificationJob) Run(_ context.Context) {
+	defer j.MarkComplete()
+
+	flags, err := evergreen.GetServiceFlags()
+	if err != nil {
+		j.AddError(errors.Wrap(err, "error retrieving admin settings"))
+
+	} else if flags == nil {
+		j.AddError(errors.Wrap(err, "fetched no service flags configuration"))
+	}
+	if len(j.NotificationID) == 0 {
+		j.AddError(errors.New("notification ID is not valid"))
+	}
+	if j.settings == nil {
+		j.settings, err = evergreen.GetConfig()
+		j.AddError(err)
+		if err == nil && j.settings == nil {
+			j.AddError(errors.New("settings object is nil"))
+		}
+	}
+	if j.HasErrors() {
+		return
+	}
+
+	n, err := notification.Find(j.NotificationID)
+	j.AddError(err)
+	if err == nil && n == nil {
+		j.AddError(errors.Errorf("can't find notification with ID: '%s", j.NotificationID))
+	}
+	if j.HasErrors() {
+		return
+	}
+
+	var sendError error
+	switch n.Subscriber.Type {
```

Comment:
I really think that this should be better encapsulated in the notification package.